File: src/api/http/project/client/routines.ts
/**
 * Routine definitions, fires, and the portable (orgtrack/v1) routine run
 * projections.
 */
import { invoke } from "@tauri-apps/api/core";

import { cachedRead, invalidateCache } from "../cache";
import type {
  RoutineDefinition,
  RoutineFire,
  RoutineFireResult,
} from "../types";

export async function listRoutines(): Promise<RoutineDefinition[]> {
  return cachedRead("__routines__:list", () => invoke("project_list_routines"));
}

export async function readRoutine(id: string): Promise<RoutineDefinition> {
  return cachedRead(`__routines__:${id}`, () =>
    invoke("project_read_routine", { id })
  );
}

export async function upsertRoutine(
  routine: RoutineDefinition
): Promise<RoutineDefinition> {
  const result = await invoke<RoutineDefinition>("project_upsert_routine", {
    routine,
  });
  invalidateCache("__routines__");
  return result;
}

export async function deleteRoutine(id: string): Promise<boolean> {
  const result = await invoke<boolean>("project_delete_routine", { id });
  invalidateCache("__routines__");
  return result;
}

export async function listRoutineFires(
  routineId: string
): Promise<RoutineFire[]> {
  return cachedRead(`__routines__:${routineId}:fires`, () =>
    invoke("project_list_routine_fires", { routineId })
  );
}

/** A row from `pm_routine_runs` (portable Routine domain, orgtrack/v1). */
export interface RoutineRunSummary {
  id: string;
  routineName: string;
  routineRevision: number;
  scopeId: string;
  status: string;
  rootWorkItemId?: string | null;
  createdBy?: string | null;
  createdAt: number;
  updatedAt: number;
}

/** Per-run projection: run row + generated WorkItems' portable states. */
export interface RoutineRunStatus {
  id: string;
  routineName: string;
  routineRevision: number;
  snapshotHash: string;
  scopeId: string;
  status: string;
  rootWorkItemId?: string | null;
  workItems: Array<{
    shortId: string;
    title: string;
    status: string;
    portableState?: string | null;
  }>;
}

/** A row from `pm_routines` (portable Routine domain, orgtrack/v1). */
export interface PortableRoutineSummary {
  name: string;
  routineId: string;
  revision: number;
  enabled: boolean;
  specHash: string;
  updatedAt: number;
}

/** List portable routines by name. Backs the Webhooks management surface. */
export async function listPortableRoutines(): Promise<
  PortableRoutineSummary[]
> {
  return invoke("project_list_portable_routines", {});
}

/** List portable routine runs, newest first. Uncached: run status moves
 *  with work-item transitions, and the surface refetches on focus. */
export async function listRoutineRuns(options?: {
  scopeId?: string;
  limit?: number;
}): Promise<RoutineRunSummary[]> {
  return invoke("project_list_routine_runs", {
    scopeId: options?.scopeId ?? null,
    limit: options?.limit,
  });
}

export async function routineRunStatus(
  runId: string
): Promise<RoutineRunStatus> {
  return invoke("project_routine_run_status", { runId });
}

export async function fireRoutine(
  routineId: string
): Promise<RoutineFireResult> {
  const result = await invoke<RoutineFireResult>("project_fire_routine", {
    routineId,
  });
  invalidateCache("__routines__");
  invalidateCache(`__routines__:${routineId}:fires`);
  return result;
}

File: src-tauri/crates/project-management/src/projects/commands/routines.rs
//! Routine commands: definitions, fire history, and materialization.

use super::super::io;
use super::super::types::{RoutineDefinition, RoutineFire};

#[tauri::command]
pub async fn project_list_routines() -> Result<Vec<RoutineDefinition>, String> {
    tokio::task::spawn_blocking(io::list_routines)
        .await
        .map_err(|err| format!("Task join error: {}", err))?
}

#[tauri::command]
pub async fn project_read_routine(id: String) -> Result<RoutineDefinition, String> {
    tokio::task::spawn_blocking(move || io::read_routine(&id))
        .await
        .map_err(|err| format!("Task join error: {}", err))?
}

#[tauri::command]
pub async fn project_upsert_routine(
    routine: RoutineDefinition,
) -> Result<RoutineDefinition, String> {
    tokio::task::spawn_blocking(move || io::upsert_routine(routine))
        .await
        .map_err(|err| format!("Task join error: {}", err))?
}

#[tauri::command]
pub async fn project_delete_routine(id: String) -> Result<bool, String> {
    tokio::task::spawn_blocking(move || io::delete_routine(&id))
        .await
        .map_err(|err| format!("Task join error: {}", err))?
}

#[tauri::command]
pub async fn project_list_routine_fires(routine_id: String) -> Result<Vec<RoutineFire>, String> {
    tokio::task::spawn_blocking(move || io::list_routine_fires(&routine_id))
        .await
        .map_err(|err| format!("Task join error: {}", err))?
}

/// List portable routines (`pm_routines`) by name. Backs the Webhooks
/// management surface; per-routine webhook state comes from
/// [`project_routine_webhook_status`].
#[tauri::command]
pub async fn project_list_portable_routines() -> Result<Vec<serde_json::Value>, String> {
    tokio::task::spawn_blocking(crate::routine_service::list_routines)
        .await
        .map_err(|err| format!("Task join error: {}", err))?
}

/// List portable routine runs (`pm_routine_runs`), newest first. Backs
/// the Runs navigation surface; per-run detail comes from
/// [`project_routine_run_status`].
#[tauri::command]
pub async fn project_list_routine_runs(
    scope_id: Option<String>,
    limit: Option<usize>,
) -> Result<Vec<serde_json::Value>, String> {
    tokio::task::spawn_blocking(move || {
        crate::routine_service::list_runs(scope_id.as_deref(), limit.unwrap_or(100))
    })
    .await
    .map_err(|err| format!("Task join error: {}", err))?
}

/// Durable run-status projection for one routine run: the run row plus
/// each generated WorkItem's portable state (orgtrack/v1 §11 ordered
/// decision procedure).
#[tauri::command]
pub async fn project_routine_run_status(run_id: String) -> Result<serde_json::Value, String> {
    tokio::task::spawn_blocking(move || crate::routine_service::run_status(&run_id))
        .await
        .map_err(|err| format!("Task join error: {}", err))?
}
